New implementations of the static check for recursive definitions

gasche opened the pull request:
This GPR is on top of #1937.

This work was conducted by Alban Reynaud @Cemoixerestre during an internship at INRIA
Saclay, with the help and under the supervision of Gabriel Scherer in
the Parsifal team.

A summary is included below. For more details, see

  https://github.com/Cemoixerestre/Internship-2018-ocaml-recursive-value

or the comments in the code itself, which has evolved a bit from
Alban's internship report above.

Summary:
--------

Rec_check is based on the idea of assigning "access modes" to variables

    m ::= Dereference (* full access and inspection of the value *)
        | Unguarded (* the value result is directly returned *)
        | Guard (* the value is stored under a data constructor *)
        | Delayed (* the value is not needed at definition time *)
        | Unused (* the value is not used at all *)

Before this patchset, the implementation (contributed by Jeremy Yallop)
was formulated as a type-checker manipulating "uses", which are mappings
from free variables to modes

    u ::= (x : m)*

and contexts Γ mapping variables to uses

    Γ ::= (x : u)*

The check relied on a judgment of the form

    Γ ⊢ e : u

where, in the typing algorithm, `Γ` and `e` are the inputs, and `u` is
the output — this could be written `-Γ ⊢ e : +u`.

After this patchset, the implementation uses a simpler notion of context
mapping variables to mods

    Γ ::= (x : m)*

and the judgment has the simpler structure

     Γ ⊢ e : m

but now `m` is an input to the judgment, indicating in which usage
context `e` is checked, and `Γ` is the output: `+Γ ⊢ e : -m`.

Other fixes:
------------

This patchset also fixes a soundness bug around `{ foo with ... }`
which is already present in 4.06 and 4.07.

Another soundness bug found during this work was submitted and fixed
independently as GPR #1915.

This work also led to the refactoring of GPR #1922, and the refactoring
of the testsuite in GPR #1937.

Commit:
New implementation of rec_check.ml (WIP)

This work was conducted during an internship at INRIA Saclay, under
the supervision of Gabriel Scherer in the Parsifal team.

A summary is included below. For more details, see

  https://github.com/Cemoixerestre/Internship-2018-ocaml-recursive-value

WIP:
----

This commit corresponds to the state of the work exactly at the end of
Alban's internship, rebased by Gabriel Scherer on trunk.
The testsuite passes, but there remain some language constructs that
are not supported by the new check implementation.

We believe that existing OCaml programs (which typically do not make
advanced use of recursive value declarations) would compile correctly
with this WIP version, but of course we plan to complete the work in
the following commits.

Summary:
--------

Rec_check is based on the idea of assigning "access modes" to variables

    m ::= Dereference | Guarded | Unguarded | Delayed

Before this commit, the implementation (contributed by Jeremy Yallop)
was formulated as a type-checker manipulating "uses", which are mappings
from free variables to modes

    u ::= (x : m)*

and contexts Γ mapping variables to uses

    Γ ::= (x : u)*

The check relied on a judgment of the form

    Γ ⊢ e : u

where, in the typing algorithm, `Γ` and `e` are the inputs, and `u` is
the output — this could be written `-Γ ⊢ e : +u`.

After this commit, the implementation uses a simpler notion of context
mapping variables to mods

    Γ ::= (x : m)*

and the judgment has the simpler structure

     Γ ⊢ e : m

but now `m` is an input to the judgment, indicating in which usage
context `e` is checked, and `Γ` is the output: `+Γ ⊢ e : -m`.

gasche committed:
rec_check: complete the missing cases from Alban's patchset

gasche committed:
rec_check: code refactoring

gasche committed:
rec_check.ml: fix a soundness bug with record override { foo with ... }

(The bug is already present in 4.06 and 4.07.)

gasche committed:
rec_check.ml: clarify the use of Guard for discarded values

gasche committed:
rec_check.ml refactoring: standard argument order for mode judgments

gasche committed:
rec_check.ml: clarify the pattern-matching rules

gasche committed:
rec_check.ml: write the term judgments in a more abstract style

We could factorize the binder judgments in a related way, but it would
require a different set of higher-order operator, and I thought that
the lower number of occurrences did not justify the extra complexity.

gasche committed:
Changes for the let-rec patchset + general commit-message explanation

This work was conducted by Alban Reynaud during an internship at INRIA
Saclay, with the help and under the supervision of Gabriel Scherer in
the Parsifal team.

A summary is included below. For more details, see

  https://github.com/Cemoixerestre/Internship-2018-ocaml-recursive-value

or the comments in the code itself, which has evolved a bit from
Alban's internship report above.

Summary:
--------

Rec_check is based on the idea of assigning "access modes" to variables

    m ::= Dereference (* full access and inspection of the value *)
        | Unguarded (* the value result is directly returned *)
        | Guard (* the value is stored under a data constructor *)
        | Delayed (* the value is not needed at definition time *)
        | Unused (* the value is not used at all *)

Before this patchset, the implementation (contributed by Jeremy Yallop)
was formulated as a type-checker manipulating "uses", which are mappings
from free variables to modes

    u ::= (x : m)*

and contexts Γ mapping variables to uses

    Γ ::= (x : u)*

The check relied on a judgment of the form

    Γ ⊢ e : u

where, in the typing algorithm, `Γ` and `e` are the inputs, and `u` is
the output — this could be written `-Γ ⊢ e : +u`.

After this patchset, the implementation uses a simpler notion of context
mapping variables to mods

    Γ ::= (x : m)*

and the judgment has the simpler structure

     Γ ⊢ e : m

but now `m` is an input to the judgment, indicating in which usage
context `e` is checked, and `Γ` is the output: `+Γ ⊢ e : -m`.

Other fixes:
------------

This patchset also fixes a soundness bug around `{ foo with ... }`
which is already present in 4.06 and 4.07.

Another soundness bug found during this work was submitted and fixed
independently as GPR #1915.

This work also led to the refactoring of GPR #1922, and the refactoring
of the testsuite in GPR #1937.

gasche commented:
(cc @yallop)

alainfrisch commented:
Do you think this might bring us closer to producing, as a side-effect of the check, a data-structure which would drive the compilation scheme and make the check "correct by construction"?

gasche commented:
I would say that this would only be marginally easier than with the previous implementation; the big conceptual jump was @yallop's idea of using a type-system-like check, this one is easier to reason about than the one it replaces, but in both cases the idea would be to look at the typed derivations as evidence.

One thing that is rather easy to do in both cases is to get a sort of dependency order on the mutual declarations. This would allow accepting more things when a topological order is just not one big cycle. But I am not sure that it helps us make the current compilation scheme "more correct".

@yallop, @Cemoixerestre and myself are thinking of writing a short paper / extended abstract on the new system, which could give me better intuitions on these things.

Thinking about compilation was on the list of things to do "if there is time" for Alban's internship, but this is a six-weeks end-of-bachelor internship, and there was no time.

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
+  (* If x is used with the mode m in e[x], and e[x] is used with mode
+     m' in e'[e[x]], then x is used with mode m'[m] in e'[e[x]].
+
+     Return is neutral for composition: m[Neutral] = m = Neutral[m].
```
`Neutral` ~> `Return` here?

gasche replied:
Good catch, I'll fix this.

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
-              (inspect (expression env e2)))
+      (*
+        G1 |- e1: m[Dereference]
+        G2 |- e2: m[Dereference] (*TODO: why is this a dereference?*)
```
There are a couple `TODO` remaining, is it on purpose?

gasche replied:
Yes, I wanted @yallop to have a chance to comment on them, because he wrote the original implementation. I believe that using `Guard` here would be correct, but I wanted his opinion before making the check more liberal than the one we replace. (Other opinions are also welcome, of course.)

yallop replied:
It's a dereference because writing a value to a mutable field involves accessing the value.  Here's a program that fails with a segmentation fault if `Dereference` is weakened to `Return` here:

```ocaml
    type t = { mutable f: float }
    let g = { f = 0.0 }
    let rec x = (g.f <- y; ()) and y = 2.0
```

(You could add the program to the test suite.)

Armael commented on testsuite/tests/letrec-check/records.ml:
```diff
@@ -0,0 +1,31 @@
+(* We have carefully placed `u` before `t` here,
+   so that the copy { t with .. }, if accepted,
+   is evaluated before 't' is initialized -- making
+   the assertion below fail, typically abording
```
abording ~> aborting ?

Armael commented on typing/rec_check.ml:
```diff
@@ -650,8 +976,9 @@ and is_destructuring_pattern : Typedtree.pattern -> bool =
+  match Env.unguarded ty idlist, Env.dependent ty idlist,
+        classify_expression expr with
   | _ :: _, _, _ (* The expression inspects rec-bound variables *)
   | _, _ :: _, Dynamic -> (* The expression depends on rec-bound variables
```
The first underscore can be turned into a `[]` which saves a few brain cycles to the reader.

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
+
+  (* Returns the most conservative mode of the two arguments.
+
+     Deref < Return < Guard < Delay < Unused
```
Deref -> Dereference

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
+    | Delay, _ -> Delay
+    | m', Return -> m'
+    | Return, m
+    | Guard, m -> m
```
This last case confused me for a moment, because I read it as `Guard[m] = m`, which is not the case.
Because of the earlier `m', Return -> m'` innocuous-looking case we have `Guard[Return] = Guard, and Guard[m] = m otherwise`. 
I wonder if it would be worth adding an explicit (redundant) case for `Guard[Return]`, to showcase more explicitly the role of the Guard mode.

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
-    (** Combine the access information of two expressions *)
+    val unguarded : t -> Ident.t list -> Ident.t list
+    (** unguarded e l: the list of all identifiers in e that are unguarded of
+        dereferenced in the environment e. *)
```
`identifiers in e` -> `identifiers in l`
Also, `ungarded of dereferenced` -> `dereferenced or returned` ?

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
-    (** Combine the access information of two expressions *)
+    val dependent : t -> Ident.t list -> Ident.t list
+    (** unguarded e l: the list of all identifiers in e that are used in e. *)
 
```
`unguarded e l` -> `dependent e l`, `identifiers in e` -> `identifiers in l`

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
-    (** No variables are accessed in an expression; it might be a
-        constant or a global identifier *)
+    val join : t -> t -> t
+    val join_list : t list -> t
```
It would maybe nice to have a comment indicating that environments sharing bindings is supported, and that in that case the right thing happens (the most conservative mode is picked).

Armael commented on typing/rec_check.ml:
```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
+
+     Deref < Return < Guard < Delay < Unused
+  *)
+  let prec m m' =
```
@gasche says a better name for this could be `join`

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+     compute the environment of a subterm from its mode,
+     so the corresponding function has type [... -> mode -> Env.t].
+
+     We write ['a -> term_judg] in this case.
```
Shouldn't `'a` be `...` instead, to match the convention used in the line above?

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+     correspond to binding constructs that have both an
+     exterior environment (the environment of the whole term)
+     and an interior environment (the environment after the binding
+     construct has introduced new names in scope).
```
Which one of Gamma and Gamma' is the exterior (resp interior) environment?

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+
+       G |- e : m + m'
+       -------------------------------
+       G |- (let x = e : m) -| x:m', G
```
There's an issue with this rule I think (name collision on `G`; and the final output of the rule is likely not just one of the two `G`s in the scope).
Would the correct rule be something like as follows?
```
       G |- e : m + m'
       -----------------------------------
       x:m', G, G' |- (let x = e : m) -| x:m', G'
```

Armael replied:
(also, I assume `m + m'` is `prec m m'`? whatever it is it should be made explicit)

gasche replied:
(done)

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+  fun f inner_mode -> fun outer_mode -> f (compos outer_mode inner_mode)
+
+(* A binding judgment [binder] expects a mode and an outer environment,
+   and returns an inner environment. [binder >> judg] computes
```
This is in contradiction with the comment above? (which says that `bind_judg` takes an interior environment and returns an exterior environment)

Armael reviewed (commented):
This is a first superficial review, I mostly read the comments but not the actual implementation yet.

Armael commented:
In the description of the PR, you give a list of 4 modes (Dereference, Unguarded, Guard, Delay), but this does not seem to match the modes before the PR (3 modes, Dereferenced, Guarded, Unguarded) or in the PR (5 modes, Dereference, Return, Guard, Delay, Unused). Could you clarify?

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+       Ge |- e : m    Gg |- g : m[Dereference]
+       G := Ge+Gg     p : mp -| G
+       ----------------------------------------
+       G - p; m[mp] |- (p (when g)? -> e) : m
```
```
...
----------
G; m |- ...
```
I assume this is an other kind of jugment? Could you explain what it means and document it in the "Usage mode judgments" comment?

gasche replied:
done, but I document each judgment right before its own recursive function

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,648 @@ let classify_expression : Typedtree.expression -> sd =
+          | Guard       otherwise
+    me := sum{G(x), x in vars(p)}
+    --------------------------------------------
+    p : (mp + me) -| G
```
Similarly, could you document this other kind of jugdment?

Cemoixerestre replied:
To answer your questions:
- the modes `Guarded` and `Dereferenced` match in the two systems. The mode `Guarded` (in the first system) is split into the three modes `Guarded`, `Delayed` and `Unused`.
- You're right, `prec m m'` and `m + m'` are the same. I should have made it explicit.

gasche replied:
done

Armael reviewed (commented):
Added a couple comments while reading the rules

gasche committed:
rec_check: clarifications suggested by Armaël

gasche commented:
I just went over @Armael's comments, improved the PR as desired and rebased. Thanks @Armael, and feel free to keep reading :-)

Armael commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+
+       G |- e : m + m'
+       -----------------------------------
+       G+G' |- (let x = e : m) -| x:m', G'
```
The closing parenthesis is off there.

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+       the evaluation of the whole program. This is the mode of
+       a variable in an expression in which it does not occur. *)
+
+  (* Returns the most conservative mode of the two arguments.
```
Minor: *more* conservative, not *most*.

I think "most conservative" could be clarified here: it means something like "the more demanding context".

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+    | Unused
+    (** [Unused] is for subexpressions that are not used at all during
+       the evaluation of the whole program. This is the mode of
+       a variable in an expression in which it does not occur. *)
```
Minor: I think `Unused` should be `Discard` or `Ignore`, since all the others are verbs.  (Alternatively, change them all to adjectives: `Dereferenced`, `Returned`, etc.)

gasche replied:
I went for `Ignore`, to me `Discard` includes the idea of an (active) action.

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+
+     In judgments we write (m + m') for (join m m').
+  *)
+  let join m m' =
```
This is just `Pervasives.min`, and I think it would be fine to define it that way (with a comment).

gasche replied:
Meh, personally I dislike making use of the barely-defined variant ordering in this way, so I will keep the current definition.

yallop replied:
Those aren't the only alternatives.  Another possibility is to define a mapping to integers:

```ocaml
  let rank = function
    | Dereference -> 0
    | Return -> 1
    | Guard -> 2
    | Delay -> 3
    | Unused -> 4
```
and then use that to define `join` in the obvious way.

This is simpler than the current definition of `join`, not least because there's no ordering in the cases.  And it turns the mode ordering (which is currently only defined in a comment) into code.

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+
+     Return is neutral for composition: m[Return] = m = Return[m].
+  *)
+  let compos m' m = match m', m with
```
Minor: `compos`?  Why not `compose`?

Aside: 'not' patterns would be useful here.

Besides noting that `Return` is neutral for composition, you might also note that `compos` is associative (I checked that it is), and that `Unused` is an annihilator.

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+      returned in the environment e. *)
+
+  val dependent : t -> Ident.t list -> Ident.t list
+  (** unguarded e l: the list of all identifiers in e that are used in e. *)
```
Two things to fix in this comment: it should say "dependent", not "unguarded", and it should say "in l", not "in e"

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+  (** An environment with no used identifiers. *)
+
+  val find : Ident.t -> t -> Mode.t
+  (** Find the mode of an indentifier in an environment.  The default mode is
```
Typo: "indentifier"

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+
+  let single id mode = M.add id mode empty
+
+  let unguarded env li =
```
Here and in `join` you define an order in the comments, and use it implicitly in the code in various places (`join`, `unguarded`, `dependent`).

Have you considered instead defining the order directly in the code (in fact, the order given by polymorphic `<` would do) , then making use of it explicitly?  For example, you could then define `unguarded` like this:

```ocaml
let unguarded env = List.filter (fun id -> lt (find id env) Guard)
```

which is clearer, since it makes use of the semantic notion of mode ordering rather than listing all the constructors.

gasche replied:
I defined a `rank : mode -> int` function to use binary ordering operators without feeling dirty about it. The code is much nicer, thanks!

I decided to change the order I used in this comment: instead of `Dereference < ... < Ignore`, the new version I just pushed uses `Ignore < ... < Dereference`¹. I think that it makes much sense (although of course the two orders are dual): we are talking about how a context "uses" a subterm, or a term uses a variable, and intuitively Ignore "uses less". 

¹: It would be fun to define this new order without changing the order of constructors in the datatype declaration, but I'm not this kind of mean person so I did change the constructor order.

yallop commented on typing/rec_check.ml:
```diff
@@ -18,145 +18,161 @@ open Types
+    in
+    List.filter (fun id -> not_guarded (find id env)) li
+
+  let dependent env li =
```
(see comment for `unguarded`)

yallop commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+
+(* Usage mode judgments.
+
+   There are two main groups of judgment functions:
```
It took me a little while to work out what was going on here.  (It would have helped if I'd remembered to look at the clearer description in the PR when reading the code, but I'm glad I didn't, since it's important that the comments in the code are at least as clear as the PR.)  I think making the comment even more comprehensive could help.  You currently say

>   - Judgments of the form "Gamma |- ... : m"
>     compute the environment of a subterm from its mode,

but it would be a little bit clearer to say 

>   - Judgments of the form "G |- e : m"
>     compute the environment G given a subterm e and mode m,

and similarly for the second form of judgement.  (Note: it's clearer to consistently use `Gamma` or `G`, and to use the names of the various components whenever you mention them.

Aside: I wonder whether there's a name for the principality property that we want the type system to have?  Given `Γ ⊢ e : τ` we say that the system has "principal types" if there's always a most general `τ` given `Γ` and `e`, and we say that it has "principal typings" if there's always a most general `Γ` and `τ` given `e`.  Here we want there to be a most general `Γ` given `e` and `τ`.

gasche replied:
What about "principal contexts"?

I believe that this form of principality is a consequence of monotonicity (if G1 < G2 and `G2 |- ..` then `G1 |- ...`) and join preservation (if `G1 | - ...` and `G2 |- ...` then `G1 + G2 |- ...`) -- note that the notation `m + m'` already assumes that two modes have a *unique* supremum. 

This system has very robust ordering properties: it is also monotonous on the `tau` side, using a weaker mode them will preserve the judgment, and compatibly with meet/min/`*` on that side (this is trivial as our order is linear), so as a consequence there is also a best/principal answer to the `G |- e : ?` question.

gasche replied:
I took your recommendations into account, and I also pushed a more comprehensive high-level documentation comment at the beginning of the file. The present comment was intended to be an implementation comment, explaining the type schemas of the functions extracted from term and binding judgments, not an explain-everything comment.

yallop commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+         (Gi; mi |- pi -> ei : m)^i
+         G |- e : sum(mi)^i
+         ----------------------------------------------
+         G + sum(Gi)^i |- match e with (pi -> ei)^i : m
```
A general comment on the ASCII typing rules: I'm glad to have them, but I find them quite hard to read, perhaps because there's so much syntax.

I don't expect everyone to share this view, but I'd find it much easier with a bit of Unicode:

         (Γᵢ; mᵢ ⊢ pᵢ → eᵢ : m)ⁱ
         Γ ⊢ e : sum(mᵢ)ⁱ
         ──────────────────────────────────────────
         Γ + sum(Γᵢ)ⁱ ⊢ match e with (pᵢ → eᵢ)ⁱ : m

gasche replied:
I could do Unicode but my understanding was that we are not currently supposed to use them (especially outside string literals), for portability/whatever reasons (cc @damiendoligez ?). I think it's better to stay ASCII-only for now.

damiendoligez replied:
Right. We haven't yet released a version of OCaml that supports Unicode source, so let's be conservative.

yallop commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+
+(* Expression judgment:
+     G |- e : m
+   where (m) is an output of the code and (G) is an output;
```
This should say "where (m) is an *input* of the code"

yallop commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+      expression e << Dereference
+    | Texp_setinstvar (pth,_,_,e) ->
+      (*
+        G |- e: m[Dereference]  (*TODO: why is this a dereference?*)
```
As with `Texp_setfield`, this really should be a `Dereference`, so we can drop the `TODO`.

Here's a program for the test suite that fails with a segmentation fault if `Dereference` is weakened to `Return`:

```ocaml
    class c = object
      val mutable f = 0.0
      method m =
        let rec x = (f <- y; ()) and y = 2.0 in f
    end
    let _ = print_float (new c)#m
```

yallop commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
-        begin match Typeopt.classify_lazy_argument e with
+      (*
+        G |- e: m[Delay]
+        ----------------  (modulo some subtle compiler optimizations)
```
I'm expecting this to break when somebody implements (e.g.) the lazy value optimization for unboxed constructors.  (But perhaps having the common code in `Typeopt` will be enough to prevent that.)

gasche replied:
Yes, I'm expecting such future changes to return one of the known-to-be-shortcut variants (for unboxed constructors, typically the classifier of the underlying argument), or to add a new variants to be handled properly here.

yallop commented on typing/rec_check.ml:
```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
-    | Tpat_alias (pat, _, _) -> is_destructuring_pattern pat
+    | Tpat_var (_id, _) -> false
+    | Tpat_alias (pat, _, _) ->
+        is_destructuring_pattern pat
```
This looks like unnecessary churn.

yallop reviewed (commented):
I think this an improvement -- thanks!  The setup feels a bit more complex, but the rules are simpler, which seems like a reasonable tradeoff.

yallop committed:
explain why field assignments may dereference the new value

gasche committed:
renamed Unused into Ignore (suggested by Jeremy Yallop)

gasche committed:
clarifications suggested by Jeremy

gasche commented:
@yallop: Thanks for the review! I did a pass over the patchset and took your comments into account.

There is one thing that we discussed with @Cemoixerestre but didn't have to work on: I believe that we can state (and prove easily by induction) meta-theorems of the following form: when `Gamma |- e : Delay`, then `Gamma` maps all the variables that do occur in `e` to `Delay`, and the others to `Ignore`. Similarly, when `Gamma |- e : Dereference`, then `Gamma` maps the occurring variables to `Dereference` and the other to `Ignore`. I plan to convince myself that these two meta-theorems do hold, and then add them in comments. The `Delay` theorem makes it easy to convince oneself that recursive functions work just fine, and the `Dereference` one gives a sort of upper bound on accepted terms.

gasche commented:
@yallop: if you believe that the PR is now in a mergeable state, it could be nice to "approve" it -- I think I could ask a maintainer to merge it then.

yallop commented:
@gasche: there are some comments that don't have replies: [here](https://github.com/ocaml/ocaml/pull/1942/files#r214454820), [here](https://github.com/ocaml/ocaml/pull/1942/files#r214455160), [here](https://github.com/ocaml/ocaml/pull/1942/files#r214455864), and [here](https://github.com/ocaml/ocaml/pull/1942/files#r214456147).  Could you please either leave a comment or mark the conversations as "resolved"?  (I'm assuming that you've simply missed them, which is unfortunately very easy to do with GitHub's interface, because one of the comments is about a typo, which should be uncontroversial.)

gasche commented:
Thanks! I missed them indeed. I will go over them now.

gasche committed:
rec_check: clarify the mode ordering, add and use a ranking function

Suggested-by: Jeremy Yallop

gasche commented:
I rebased the PR with a more substantial high-level comment at the top of the file, which I would like to be at least as clear as the summary of this PR, without the historical aspect (old and new system) which is not relevant in the versioned codebase.

gasche committed:
rec_check.ml: high-level documentation and reordering

We provide an ocamldoc-style module comment to explain the general
lines of the recursive check. This complements existing comments about
the term- and binding-judgments that are more implementation-oriented.

It was natural to reorder the code so that the "static or dynamic
size" code is apart from the "usage of recursive variables" code.

Finally, I also added copyright headers to reflect the now shared
authorship of the code. (Alban has signed a CLA.)

yallop commented:
> I rebased the PR with a more substantial high-level comment at the top of the file,

Thanks.  I think the new comment is a very useful addition.

yallop commented:
(I'll approve once the bug that's causing Travis to fail is fixed.)

gasche commented:
Duh... this should be fixed now.

gasche commented:
(just fixed the `check-typo` check)

damiendoligez commented on typing/rec_check.ml:
```diff
@@ -3,6 +3,8 @@
+(*               Gabriel Scherer, Project Parsifal, INRIA Saclay          *)
+(*               Alban Reynaud, ENS Lyon                                  *)
 (*                                                                        *)
 (*   Copyright 2017 Jeremy Yallop                                         *)
```
You should add a copyright line for INRIA and for either Alban (if the internship was a mandatory part of his cursus) or ENS Lyon (if the internship was optional).

damiendoligez commented on typing/rec_check.ml:
```diff
@@ -12,151 +14,97 @@
+address. This is a case of productive/guarded recursion.
+
+On the contrary, [let rec x = x] is unguarded recursion (the meaning
+is undertermined), and [let rec x = x+1] would need the value of [x]
```
typo "undetermined"

damiendoligez commented on typing/rec_check.ml:
```diff
@@ -306,333 +252,841 @@ let classify_expression : Typedtree.expression -> sd =
+    | Guard, Return -> Guard
+    | Guard, ((Dereference | Guard | Delay) as m) -> m
+    | Return, Return -> Return
+    | Return, ((Dereference | Guard | Delay) as m) -> m
```
Why not just write `| Return, m -> m` instead of these two lines? It would be clearer that it does what the comment says.

gasche replied:
I want to highlight the subtle difference between `Guard` and `Return` in how they compose, by making the definitions look similar. We originally had a more compact version, namely

```ocaml
    | m', Return -> m'
    | Return, m
    | Guard, m -> m
```

and @Armael pointed out that it was hard to reason about.

The current definition uses `as m` to point out that the cases are treated homogeneously, *and* lists the case explicitly to help readers reason about which specific cases can flow to these clauses. Best of both worlds!

damiendoligez commented on typing/rec_check.ml:
```diff
@@ -306,333 +252,841 @@ let classify_expression : Typedtree.expression -> sd =
-              (inspect (option expression env eo)))
+      (*
+        G |- e: m[Dereference]
+        ---------------------- (plus weird 'eo' option)
```
"weird"? That word doesn't make you look confident...

gasche replied:
It doesn't, because we aren't. The code must be safe because `Dereference` is the most conservative possible choice for `eo`, but I still have mostly no idea what this `eo` thing does (despite having looked at the compilation scheme). It's right that people hitting this case would be less-confident and would want (if they are changing it) to check things by themselves.

yallop replied:
> but I still have mostly no idea what this `eo` thing does

I'm not sure how much "mostly" covers here, but the `eo` is for representing calls to a superclass method through an ancestor variable, like this:

```ocaml
class b = object
  method m = 0
end
class d = object
  inherit b as parent   (* 'parent' is an ancestor *)
  method m = parent#m  (* Call to method 'm' in class 'b' *)
end
```

damiendoligez reviewed (approved):
Approving based on @yallop's review.

I have a few superficial remarks.

gasche commented:
@damiendoligez Thanks! I took the comments into account.

gasche committed:
[minor] add rec_check.ml copyright info (as suggested by Damien)
